OSDOCS-4928: Docs for cert-manager operator GA (#56754)
openshift/openshift-docs: modules/cert-manager-acme-dns01-aws.adoc

* You have set up an IAM role for Amazon Route 53. For more information, see link:https://cert-manager.io/docs/configuration/acme/dns01/route53/[Route53] in the upstream cert-manager documentation.
+
[NOTE]
====
For the {cert-manager-operator}, you must provide an `accessKeyID` and `secretAccessKey` to specify the credentials.
====

.Procedure

. Configure the nameserver settings for the DNS01 self check:

.. Edit the `CertManager` resource by running the following command:
+
[source,terminal]
----
$ oc edit certmanager cluster
----

.. Add a `spec.controllerConfig` section with the following override arguments:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
metadata:
  name: cluster
  ...
spec:
  ...
  controllerConfig:                                <1>
    overrideArgs:
      - '--dns01-recursive-nameservers=1.1.1.1:53' <2>
      - '--dns01-recursive-nameservers-only'       <3>
----
<1> Add the `spec.controllerConfig` section.
<2> Provide a comma-separated list of `<host>:<port>` nameservers to query for the DNS01 self check.
<3> Specify to only use recursive nameservers.
+

> // TODO: what exactly does this mean? That it uses caching instead of checking the authoritative nameservers every time? Might cause it to take longer it says why? (every time)?

Author: @thejasn @swghosh Can you clarify what this "--dns01-recursive-nameservers-only" really means? The upstream docs weren't clear to me.

Reviewer: It basically forces cert-manager to not search for authoritative nameservers (associated with that domain) at all and only use the recursive nameservers that  are provided. The part where it say this might take longer due to caches, even I'm not very clear on. Will try to get back on this.

Author: Yeah it seemed like if it was caching it, then it should actually take less time. I can leave this information out for now, but this would be good to state if we know one way or the other.

Author: Not worrying about this for now. If you find that we should say anything about time it takes with vs without this option, let me know and we can add later.

+
// TODO Find out if that is just an example value, where they need to replace?

.. Save the file to apply the changes.

. Create an issuer.

.. Create a YAML file that defines the `Issuer` object:
+
.Example `issuer.yaml` file
[source,yaml]
----
  apiVersion: cert-manager.io/v1
  kind: Issuer
  metadata:
    name: letsencrypt-staging                                        <1>
    namespace: sandbox                                               <2>
  spec:
    acme:
      server: https://acme-staging-v02.api.letsencrypt.org/directory <3>
      email: "thn@redhat.com"                                        <4>
      privateKeySecretRef:
        name: letsencrypt-staging                                    <5>
      solvers:
      - dns01:
          route53:
            accessKeyID: AKIAT5TK4ZONNAZQMGOD                        <6>
            hostedZoneID: Z3URY6TWQ91KVV                             <7>
            region: us-east-1
            secretAccessKeySecretRef:
              name: "aws-secret"                                     <8>
              key: "awsSecretAccessKey"                              <9>
        selector:
          dnsNames:

>           - 'example.org'                                            <10>

Reviewer: Since most of the fields here are explained in upstream docs, we don't need to go into too much detail explaining each field. We can just mention what fields need to be filled in by the user and just add a reference to the upstream API reference.

----
<1> TODO

> <2> Specify the namespace for the CA issuer.

Author: @thejasn @swghosh The example I copied from used "sandbox" as the example namespace. Two questions:

1. Do we need a step in here to create that namespace first?
2. Any suggestions on a better example namespace to use other than sandbox?

Reviewer: > Do we need a step in here to create that namespace first?
Yes, you can add it as a prerequisite or have an explicit step with `oc new-project sandbox`

> Any suggestions on a better example namespace to use other than sandbox?
Honestly the namespace isn't very significant, I'm ok with it being sandbox or if there is some sort of convention in OCP docs we can go with that as well.

> <3> TODO

Author: @thejasn @swghosh Does the `https://acme-staging-v02.api.letsencrypt.org/directory` value ever change, or should they always use this exact value?

Can you give me a description of what this is?

Reviewer: https://cert-manager.io/docs/reference/api-docs/#acme.cert-manager.io/v1.ACMEIssuer

> 	Server is the URL used to access the ACME server’s ‘directory’ endpoint. For example, for Let’s Encrypt’s staging endpoint, you would use: “https://acme-staging-v02.api.letsencrypt.org/directory”. Only ACME v2 endpoints (i.e. RFC 8555) are supported.

Author: @thejasn So they would use this if they're using Let's Encrypt, but could change this to something else if they were using something other than Let's Encrypt, correct?

Reviewer: Not necessarily, letsencrypt could have other server URLs as well that could be provided here. The staging URL is just a valid sample ACME server URL, the docs could have this sample or a complete dummy URL (both of which needs to be replaced by the user).

<4> #❗Replace this with your own email address

> <5> TODO

Author: @thejasn @swghosh What is this `privateKeySecretRef` of `letsencrypt-staging` referencing? I don't see this anywhere else in the procedure. I need to know what to say that this should be refereencing

Reviewer: > 	privateKeySecretRef is the name of a Kubernetes Secret resource that will be used to store the automatically generated ACME account private key. Optionally, a key may be specified to select a specific entry within the named Secret resource. If key is not specified, a default of tls.key will be used.

https://cert-manager.io/docs/reference/api-docs/#acme.cert-manager.io/v1.ACMEIssuer

Author: @thejasn I'm assuming that his has to exist already then. Should we add a step to add this secret? Or just say that it has to already exist?

Reviewer: Works both ways, we can recommend to rely it creating a new secret rather than update an existing secret (at least for a fresh setup).

<6> # replace with your aws key ID
<7>  # replace with your hosted zone ID

> <8> TODO

Reviewer: These are user defined values. 

https://cert-manager.io/docs/reference/api-docs/#acme.cert-manager.io/v1.ACMEIssuerDNS01ProviderRoute53

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-operator-issuer-acme.adoc

:_content-type: PROCEDURE
[id="cert-manager-acme-dns01-aws_{context}"]
= Configuring an ACME issuer to solve DNS-01 challenges

You can use {cert-manager-operator} to set up an ACME issuer to solve DNS-01 challenges. This procedure uses _Let's Encrypt_ as the ACME CA server and shows how to solve DNS-01 challenges with Amazon Route 53.

[NOTE]
====
It is not supported to use private DNS zones.
====

.Prerequisites

* You have access to the {product-title} cluster as a user with the `cluster-admin` role.
* You have set up an IAM role for Amazon Route 53. For more information, see link:https://cert-manager.io/docs/configuration/acme/dns01/route53/[Route53] in the upstream cert-manager documentation.
+
[NOTE]
====

> For the {cert-manager-operator}, you must provide an `accessKeyID` and `secretAccessKey` to specify the credentials.

Reviewer 2: This is not fully true. CC @anjaltelang @thejasn . I was not sure of this until yesterday on the weekend I had enough time to finally investigate https://cert-manager.io/docs/configuration/acme/dns01/route53/#eks-iam-role-for-service-accounts-irsa and successfully tried in OCP AWS STS env https://docs.openshift.com/container-platform/4.12/authentication/managing_cloud_provider_credentials/cco-mode-sts.html ; in OCP AWS STS env, `accessKeyID` and `secretAccessKey` are not a "must". Let me add my steps and results I tried in my test doc.
```suggestion
For the {cert-manager-operator}, you must provide explicit `accessKeyID` and `secretAccessKey` to specify the credentials unless your cluster is configured to use AWS STS, in which case you can use implicit ambient credentials (e.g. inferred from the environment).
```

Reviewer: We can definitely call out that these steps are for non STS clusters.

Author: @thejasn Which of the steps in this procedure are for non-STS clusters only?

Reviewer: Pretty much all steps across the board were given for non STS clusters only (some may just work on STS but that was not what we had in mind when we gave the steps). We can pull in @xingxingxia's suggestion.

Reviewer 2: Agree

Author: Thanks, reworded some. Let me know if it looks okay:

```
If your cluster is _not_ configured to use the AWS Security Token Service (STS), you must must provide explicit `accessKeyID` and `secretAccessKey` credentials. If you cluster uses AWS STS, you can use implicit ambient credentials.
```

> If your cluster is _not_ configured to use the AWS Security Token Service (STS), you must must provide explicit `accessKeyID` and `secretAccessKey` credentials. If you cluster uses AWS STS, you can use implicit ambient credentials.

Reviewer 2: > you must must provide

Duplicate "must"

Author: Good catch, updated!

====

.Procedure

. Optional: Override the nameserver settings for the DNS-01 self check:
+

> This is only required when the target public-hosted zone overlaps with the cluster's default private-hosted zone.

Reviewer 3: ```suggestion
This step is required only when the target public-hosted zone overlaps with the cluster's default private-hosted zone.
```

.. Edit the `CertManager` resource by running the following command:
+
[source,terminal]
----
$ oc edit certmanager cluster
----

.. Add a `spec.controllerConfig` section with the following override arguments:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
metadata:
  name: cluster
  ...
spec:
  ...
  controllerConfig:                                <1>
    overrideArgs:
      - '--dns01-recursive-nameservers=1.1.1.1:53' <2>
      - '--dns01-recursive-nameservers-only'       <3>
----
<1> Add the `spec.controllerConfig` section.
<2> Provide a comma-separated list of `<host>:<port>` nameservers to query for the DNS-01 self check.

> <3> Specify to only use recursive nameservers instead of checking the authoritative nameservers associated with that domain.

Reviewer 3: Check the placement of "only" in <3>. I'm not sure if "only" should modify "use" or "recursive nameservers". See https://www.ibm.com/docs/en/ibm-style?topic=grammar-adverbs-only.

    -n my-issuer-namespace
----
<1> Replace `<aws_secret_access_key>` with your AWS secret access key.

. Create an issuer.

.. Create a YAML file that defines the `Issuer` object:
+
.Example `issuer.yaml` file
[source,yaml]
----
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt-staging                                        <1>
  namespace: my-issuer-namespace                                   <2>
spec:
  acme:
    server: https://acme-staging-v02.api.letsencrypt.org/directory <3>
    email: "<email_address>"                                       <4>
    privateKeySecretRef:
      name: <secret_private_key>                                   <5>
    solvers:
    - dns01:
        route53:
          accessKeyID: <aws_key_id>                                <6>
          hostedZoneID: <hosted_zone_id>                           <7>
          region: us-east-1
          secretAccessKeySecretRef:
            name: "aws-secret"                                     <8>
            key: "awsSecretAccessKey"                              <9>
----
<1> Provide a name for the issuer.
<2> Specify the namespace that you created for the issuer.

> <3> Specify the URL to access the ACME server's `directory` endpoint. This example uses the _Let's Encrypt_ staging environment.

Reviewer 3: See my earlier comment about whether to italicize "Let's Encrypt".

<4> Replace `<email_address>` with your email address.
<5> Replace `<secret_private_key>` with the name of the secret to store the ACME account private key in.
<6> Replace `<aws_key_id>` with your AWS key ID.
<7> Replace `<hosted_zone_id>` with your hosted zone ID.
<8> Specify the name of the secret you created.
<9> Specify the key in the secret you created that stores your AWS secret access key.

.. Create the `Issuer` object by running the following command:
+
[source,terminal]
----
$ oc create -f issuer.yaml
----

. Create a certificate.

.. Create a YAML file that defines the `Certificate` object:
+
.Example `certificate.yaml` file
[source,yaml]
----
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: my-tls-cert              <1>
  namespace: my-issuer-namespace <2>
spec:
  isCA: false
  commonName: '<common_name>'    <3>
  secretName: my-tls-cert        <4>
  dnsNames:
  - '<domain_name>'              <5>
  issuerRef:
    name: letsencrypt-staging    <6>
    kind: Issuer
----
<1> Provide a name for the certificate.
<2> Specify the namespace that you created for the issuer.
<3> Replace `<common_name>` with your common name (CN).

> <4> The name of the secret to create that will contain the certificate.

Reviewer 3: Should <4> start with "Replace" or "Specify" as the other notes do? For example, "Specify the name of the secret to create that will contain the certificate."

Author: Updating!

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-operator-issuer-acme.adoc

:_content-type: PROCEDURE
[id="cert-manager-acme-dns01-aws_{context}"]
= Configuring an ACME issuer to solve DNS-01 challenges

You can use {cert-manager-operator} to set up an ACME issuer to solve DNS-01 challenges. This procedure uses _Let's Encrypt_ as the ACME CA server and shows how to solve DNS-01 challenges with Amazon Route 53.

.Prerequisites

* You have access to the {product-title} cluster as a user with the `cluster-admin` role.
* You have set up an IAM role for Amazon Route 53. For more information, see link:https://cert-manager.io/docs/configuration/acme/dns01/route53/[Route53] in the upstream cert-manager documentation.
+
[NOTE]
====
For the {cert-manager-operator}, you must provide an `accessKeyID` and `secretAccessKey` to specify the credentials.
====

.Procedure

. Configure the nameserver settings for the DNS-01 self check:

.. Edit the `CertManager` resource by running the following command:
+
[source,terminal]
----
$ oc edit certmanager cluster
----

.. Add a `spec.controllerConfig` section with the following override arguments:

> +

Reviewer 2: Saw @thejasn talks about the 2 args --dns01-recursive-nameservers and ...-only in https://github.com/openshift/openshift-docs/pull/56754#issuecomment-1469517451 . But for me, when I tested Route 53 in AWS and GCP clusters, I needed not set. See my verification steps in https://issues.redhat.com/browse/CM-11, I wrote there. And for dns01 other providers Azure DNS and Google Cloud DNS, I needed not set the 2 flags, either. I think they are not needed, but you can document they may be optional.

Reviewer: @bergerhoffer hold off on making changes to this section for now. We are having a parallel discussion around testing this. Will update here once we have a conclusion.

Author: @thejasn @xingxingxia Is there any conclusion to this? Posted some followup questions in the slack thread

Author: Updating to say that Anjali answered in slack, so I've updated per that

----
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
metadata:
  name: cluster
  ...
spec:
  ...
  controllerConfig:                                <1>
    overrideArgs:
      - '--dns01-recursive-nameservers=1.1.1.1:53' <2>
      - '--dns01-recursive-nameservers-only'       <3>
----
<1> Add the `spec.controllerConfig` section.
<2> Provide a comma-separated list of `<host>:<port>` nameservers to query for the DNS-01 self check.
<3> Specify to only use recursive nameservers instead of checking the authoritative nameservers associated with that domain.
+
// TODO: Upstream docs say that this might take longer using this - determine if that's true/state here if so in a future update

.. Save the file to apply the changes.

. Create a namespace for the issuer.

.. Create a YAML file that defines a `Namespace` object:
+
.Example `namespace.yaml` file
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: my-issuer-namespace <1>
----
<1> Specify the namespace for the issuer.

.. Create the `Namespace` object by running the following command:
+
[source,terminal]
----
$ oc create -f namespace.yaml
----

. Create a secret to store your AWS credentials in.

> .. Create a YAML file that defines a `Secret` object:
> +
> .Example `secret.yaml` file
> [source,yaml]
> ----
> apiVersion: v1
> kind: Secret
> metadata:
>   name: "aws-secret"                          <1>
>   namespace: my-issuer-namespace              <2>
> type: Opaque
> data:
>   awsSecretAccessKey: <aws_secret_access_key> <3>
> ----
> <1> Specify the name for the secret.
> <2> Specify the namespace for the secret.
> <3> Replace `<aws_secret_access_key>` with your AWS secret access key.
> 
> .. Create the `Secret` object by running the following command:
> +
> [source,terminal]
> ----
> $ oc create -f secret.yaml
> ----

Reviewer 2: > <3> Replace `<aws_secret_access_key>` with your AWS secret access key.

This statement is incorrect. Within a "Secret" resource, the field's value will be the base64-encoded value of the access key. You will have to document the base64 command if needed for readers. For simplicity, a command is better than above many lines:
```suggestion
[source,terminal]
----
$ oc create secret generic aws-secret --from-literal=awsSecretAccessKey=<aws_secret_access_key> -n my-issuer-namespace
----
```

Author: @xingxingxia Thanks, I updated to use the command. With using the command, if I use that callout now, would that be correct? (I'm pushing an update that will use it, but let me know if I need to say something different).

. Create an issuer.

.. Create a YAML file that defines the `Issuer` object:
+
.Example `issuer.yaml` file
[source,yaml]
----
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt-staging                                        <1>
  namespace: my-issuer-namespace                                   <2>
spec:
  acme:
    server: https://acme-staging-v02.api.letsencrypt.org/directory <3>
    email: "<email_address>"                                       <4>
    privateKeySecretRef:
      name: <secret_private_key>                                   <5>
    solvers:
    - dns01:
        route53:
          accessKeyID: <aws_key_id>                                <6>
          hostedZoneID: <hosted_zone_id>                           <7>
          region: us-east-1
          secretAccessKeySecretRef:
            name: "aws-secret"                                     <8>
            key: "awsSecretAccessKey"                              <9>
      selector:
        dnsNames:

>         - '<domain_name>'                                          <10>

Reviewer 2: @bergerhoffer you can remove the "selector" stanza, it is not needed for our document.
"dnsNames" under "selector" require an exact match (see https://cert-manager.io/docs/configuration/acme/#dns-names says "dnsNames take an exact match") and the usage of "selector" field actually occupies significant documenting: after "The solvers stanza has an optional selector field" in https://cert-manager.io/docs/configuration/acme/ , https://cert-manager.io/docs/configuration/acme/ uses half page for "selector" field and a standalone section for "dnsNames". The function of "dnsNames" in "Issuer" yaml is not the same function of "dnsNames" field in Certificate. If you want to understand more, you can look at my https://docs.google.com/document/d/1ZUzGBIZVOwqQrEB-M0zE9aAjK8_ue7Gfsxv_jm4XX7Q/ dns01 examples.

<5> Replace `<secret_private_key>` with the name of the secret to store the ACME account private key in.
<6> Replace `<aws_key_id>` with your AWS key ID.
<7> Replace `<hosted_zone_id>` with your hosted zone ID.
<8> Specify the name of the secret you created.
<9> Specify the key in the secret you created that stores your AWS secret access key.
<10> Replace `<domain_name>` with your domain name.

.. Create the `Issuer` object by running the following command:
+
[source,terminal]
----
$ oc create -f issuer.yaml
----

. Create a certificate.

.. Create a YAML file that defines the `Certificate` object:
+
.Example `certificate.yaml` file
[source,yaml]
----
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: my-tls-cert              <1>
  namespace: my-issuer-namespace <2>
spec:
  isCA: false
  commonName: '<common_name>'    <3>
  secretName: my-tls-cert        <4>
  dnsNames:
  - '<domain_name>'              <5>
  issuerRef:
    name: letsencrypt-staging    <6>
    kind: Issuer
----
<1> Provide a name for the certificate.
<2> Specify the namespace that you created for the issuer.
<3> Replace `<common_name>` with your common name (CN).

> <4> TODO: not sure what this is representing

Reviewer: The secret that will be created which will contain the certificate.
